File: app/src/main/java/com/hotsharefile/hotsharefile/QR.java
package com.hotsharefile.hotsharefile;

import android.graphics.Bitmap;
import android.graphics.Color;

import java.nio.charset.StandardCharsets;

public class QR {

    public static Bitmap GenerateQrCode(int targetSize, String text) {
        QrEncoder encoder = new QrEncoder();
        return encoder.encode(targetSize, text);
    }

    private static class QrEncoder {

        // Precomputed Galois Field tables for Reed-Solomon error correction.
        // Made static to avoid recalculation on every QR generation.
        private static final int[] EXP = new int[512];
        private static final int[] LOG = new int[256];

        static {
            int x = 1;
            for (int i = 0; i < 255; i++) {
                EXP[i] = x;
                LOG[x] = i;
                x <<= 1;
                if ((x & 0x100) != 0) x ^= 0x11D; // QR Code primitive polynomial
            }
            for (int i = 255; i < 512; i++) {
                EXP[i] = EXP[i - 255];
            }
        }

        private int gfMultiply(int x, int y) {
            if (x == 0 || y == 0) return 0;
            return EXP[LOG[x] + LOG[y]];
        }

        private int[] getGenerator(int degree) {
            int[] gen = {1};
            for (int i = 0; i < degree; i++) {
                int[] nextGen = new int[gen.length + 1];
                nextGen[0] = gen[0];
                for (int j = 1; j < gen.length; j++) {
                    nextGen[j] = gen[j] ^ gfMultiply(gen[j - 1], EXP[i]);
                }
                nextGen[gen.length] = gfMultiply(gen[gen.length - 1], EXP[i]);
                gen = nextGen;
            }
            return gen;
        }

        private int[] calculateEcc(int[] data, int[] gen) {
            int[] msg = new int[data.length + gen.length - 1];
            System.arraycopy(data, 0, msg, 0, data.length);
            for (int i = 0; i < data.length; i++) {
                int factor = msg[i];
                if (factor == 0) continue;
                for (int j = 0; j < gen.length; j++) {
                    msg[i + j] ^= gfMultiply(gen[j], factor);
                }
            }
            int[] ecc = new int[gen.length - 1];
            System.arraycopy(msg, data.length, ecc, 0, ecc.length);
            return ecc;
        }

        private void drawFinder(int[][] matrix, int startX, int startY) {
            for (int y = -1; y <= 7; y++) {
                for (int x = -1; x <= 7; x++) {
                    int cx = startX + x;
                    int cy = startY + y;
                    if (cx < 0 || cx >= 33 || cy < 0 || cy >= 33) continue;

                    if (x == -1 || x == 7 || y == -1 || y == 7) {
                        matrix[cy][cx] = 0;
                    } else if (x == 0 || x == 6 || y == 0 || y == 6) {
                        matrix[cy][cx] = 1;
                    } else if (x == 1 || x == 5 || y == 1 || y == 5) {
                        matrix[cy][cx] = 0;
                    } else {
                        matrix[cy][cx] = 1;
                    }
                }
            }
        }

        public Bitmap encode(int tSize, String content) {
            byte[] rawData = content.getBytes(StandardCharsets.UTF_8);
            if (rawData.length > 62) {
                throw new IllegalArgumentException("Text exceeds Version 4-M capacity (62 bytes).");
            }

            boolean[] bits = new boolean[512];
            int bitPos = 0;

            // Byte Mode Indicator (0100)
            bits[bitPos++] = false; bits[bitPos++] = true;
            bits[bitPos++] = false; bits[bitPos++] = false;

            // Length Indicator (8 bits for Version 4)
            int len = rawData.length;
            for (int i = 7; i >= 0; i--) {
                bits[bitPos++] = ((len >> i) & 1) == 1;
            }

            // Data Bits
            for (byte b : rawData) {
                for (int i = 7; i >= 0; i--) {
                    bits[bitPos++] = ((b >> i) & 1) == 1;
                }
            }

            // Terminator (Up to 4 zeros)
            int terminatorLimit = Math.min(4, 512 - bitPos);
            for (int i = 0; i < terminatorLimit; i++) {
                bits[bitPos++] = false;
            }

            // Byte Alignment Padding
            while (bitPos % 8 != 0) {
                bits[bitPos++] = false;
            }

            // Padding Bytes (0xEC and 0x11)
            boolean padSwitch = true;
            while (bitPos < 512) {
                int padByte = padSwitch ? 0xEC : 0x11;
                for (int i = 7; i >= 0; i--) {
                    bits[bitPos++] = ((padByte >> i) & 1) == 1;
                }
                padSwitch = !padSwitch;
            }

            // Extract Bytes
            int[] dataWords = new int[64];
            for (int i = 0; i < 64; i++) {
                int b = 0;
                for (int j = 0; j < 8; j++) {
                    if (bits[i * 8 + j]) b |= (1 << (7 - j));
                }
                dataWords[i] = b;
            }

            // Split into 2 blocks for Version 4-M
            int[] block1Data = new int[32];
            int[] block2Data = new int[32];
            System.arraycopy(dataWords, 0, block1Data, 0, 32);
            System.arraycopy(dataWords, 32, block2Data, 0, 32);

            int[] gen = getGenerator(18);
            int[] block1Ecc = calculateEcc(block1Data, gen);
            int[] block2Ecc = calculateEcc(block2Data, gen);

            // Interleave Data and ECC
            int[] finalData = new int[100];
            int idx = 0;
            for (int i = 0; i < 32; i++) {
                finalData[idx++] = block1Data[i];
                finalData[idx++] = block2Data[i];
            }
            for (int i = 0; i < 18; i++) {
                finalData[idx++] = block1Ecc[i];
                finalData[idx++] = block2Ecc[i];
            }

            int[][] matrix = new int[33][33];
            for (int y = 0; y < 33; y++) {
                for (int x = 0; x < 33; x++) matrix[y][x] = -1;
            }

            // Patterns
            drawFinder(matrix, 0, 0);
            drawFinder(matrix, 26, 0);
            drawFinder(matrix, 0, 26);

            // Alignment Pattern for Version 4 at (26, 26)
            for (int y = -2; y <= 2; y++) {
                for (int x = -2; x <= 2; x++) {
                    boolean isBlack = (x == -2 || x == 2 || y == -2 || y == 2 || (x == 0 && y == 0));
                    matrix[26 + y][26 + x] = isBlack ? 1 : 0;
                }
            }

            // Timing Patterns
            for (int i = 8; i < 25; i++) {
                matrix[6][i] = (i % 2 == 0) ? 1 : 0;
                matrix[i][6] = (i % 2 == 0) ? 1 : 0;
            }

            // Dark Module
            matrix[25][8] = 1;

            // Reserve Format Info Area
            for (int i = 0; i < 9; i++) {
                if (matrix[8][i] == -1) matrix[8][i] = 2;
                if (matrix[i][8] == -1) matrix[i][8] = 2;
            }
            for (int i = 25; i < 33; i++) {
                if (matrix[8][i] == -1) matrix[8][i] = 2;
                if (matrix[i][8] == -1) matrix[i][8] = 2;
            }

            // Data Placement (ZigZag)
            int bitIdx = 0;
            int dir = -1;
            int x = 32;
            int y = 32;
            while (x > 0) {
                if (x == 6) x--;
                while (y >= 0 && y < 33) {
                    for (int i = 0; i < 2; i++) {
                        int cx = x - i;
                        if (matrix[y][cx] == -1) {
                            boolean bit = false;
                            if (bitIdx < 800) {
                                bit = ((finalData[bitIdx / 8] >> (7 - (bitIdx % 8))) & 1) == 1;
                            }
                            // Apply Mask 0: (row + col) % 2 == 0
                            boolean maskBit = ((y + cx) % 2 == 0);
                            matrix[y][cx] = (bit ^ maskBit) ? 1 : 0;
                            bitIdx++;
                        }
                    }
                    y += dir;
                }
                dir = -dir;
                y += dir;
                x -= 2;
            }

            // Format Info Bits Hardcoded (Level M, Mask 0 -> 0x5412)
            // LSB -> MSB sequence exact standard match
            int[] formatBits = {0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1};

            int[] xs1 = {8, 8, 8, 8, 8, 8, 8, 8, 7, 5, 4, 3, 2, 1, 0};
            int[] ys1 = {0, 1, 2, 3, 4, 5, 7, 8, 8, 8, 8, 8, 8, 8, 8};

            int[] xs2 = {8, 8, 8, 8, 8, 8, 8, 25, 26, 27, 28, 29, 30, 31, 32};
            int[] ys2 = {32, 31, 30, 29, 28, 27, 26, 8, 8, 8, 8, 8, 8, 8, 8};

            for (int i = 0; i < 15; i++) {
                matrix[ys1[i]][xs1[i]] = formatBits[i];
                matrix[ys2[i]][xs2[i]] = formatBits[i];
            }

            // --- PERFECT PIXEL RENDERING (NO CANVAS FLOAT BLUR) ---
            int margin = 4;
            int matrixSize = 33;
            int totalModules = matrixSize + (2 * margin);

            // Calculate an exact integer scale to avoid sub-pixel blurring
            int scale = Math.max(1, tSize / totalModules);
            int finalPixelSize = scale * totalModules;

            int[] pixels = new int[finalPixelSize * finalPixelSize];
            for (int py = 0; py < finalPixelSize; py++) {
                for (int px = 0; px < finalPixelSize; px++) {
                    int moduleY = (py / scale) - margin;
                    int moduleX = (px / scale) - margin;

                    boolean isBlack = false;
                    if (moduleY >= 0 && moduleY < matrixSize && moduleX >= 0 && moduleX < matrixSize) {
                        if (matrix[moduleY][moduleX] == 1) {
                            isBlack = true;
                        }
                    }
                    pixels[py * finalPixelSize + px] = isBlack ? Color.BLACK : Color.WHITE;
                }
            }

            // Create Bitmap directly from pixel array (Maximum sharpness)
            Bitmap bmp = Bitmap.createBitmap(finalPixelSize, finalPixelSize, Bitmap.Config.ARGB_8888);
            bmp.setPixels(pixels, 0, finalPixelSize, 0, 0, finalPixelSize, finalPixelSize);

            return bmp;
        }
    }
}

File: app/src/main/java/com/hotsharefile/hotsharefile/QR.kt
package com.hotsharefile.hotsharefile

import android.graphics.Bitmap
import android.graphics.Color
import java.nio.charset.StandardCharsets
import kotlin.math.max
import kotlin.math.min

object QR {

    fun generateQrCode(targetSize: Int, text: String): Bitmap {
        return QrEncoder.encode(targetSize, text)
    }

    private object QrEncoder {

        private val EXP = IntArray(512)
        private val LOG = IntArray(256)

        init {
            var x = 1
            for (i in 0 until 255) {
                EXP[i] = x
                LOG[x] = i
                x = x shl 1
                if ((x and 0x100) != 0) x = x xor 0x11D
            }
            for (i in 255 until 512) {
                EXP[i] = EXP[i - 255]
            }
        }

        private fun gfMultiply(x: Int, y: Int): Int {
            if (x == 0 || y == 0) return 0
            return EXP[LOG[x] + LOG[y]]
        }

        private fun getGenerator(degree: Int): IntArray {
            var gen = intArrayOf(1)
            for (i in 0 until degree) {
                val nextGen = IntArray(gen.size + 1)
                nextGen[0] = gen[0]
                for (j in 1 until gen.size) {
                    nextGen[j] = gen[j] xor gfMultiply(gen[j - 1], EXP[i])
                }
                nextGen[gen.size] = gfMultiply(gen[gen.size - 1], EXP[i])
                gen = nextGen
            }
            return gen
        }

        private fun calculateEcc(data: IntArray, gen: IntArray): IntArray {
            val msg = IntArray(data.size + gen.size - 1)
            System.arraycopy(data, 0, msg, 0, data.size)
            for (i in data.indices) {
                val factor = msg[i]
                if (factor == 0) continue
                for (j in gen.indices) {
                    msg[i + j] = msg[i + j] xor gfMultiply(gen[j], factor)
                }
            }
            val ecc = IntArray(gen.size - 1)
            System.arraycopy(msg, data.size, ecc, 0, ecc.size)
            return ecc
        }

        private fun drawFinder(matrix: Array<IntArray>, startX: Int, startY: Int) {
            for (y in -1..7) {
                for (x in -1..7) {
                    val cx = startX + x
                    val cy = startY + y
                    if (cx !in 0..32 || cy !in 0..32) continue

                    matrix[cy][cx] = when {
                        x == -1 || x == 7 || y == -1 || y == 7 -> 0
                        x == 0 || x == 6 || y == 0 || y == 6 -> 1
                        x == 1 || x == 5 || y == 1 || y == 5 -> 0
                        else -> 1
                    }
                }
            }
        }

        fun encode(tSize: Int, content: String): Bitmap {
            val rawData = content.toByteArray(StandardCharsets.UTF_8)
            require(rawData.size <= 62) { "Text exceeds Version 4-M capacity (62 bytes)." }

            val bits = BooleanArray(512)
            var bitPos = 0

            // Byte Mode Indicator (0100)
            bits[bitPos++] = false; bits[bitPos++] = true
            bits[bitPos++] = false; bits[bitPos++] = false

            // Length Indicator (8 bits for Version 4)
            val len = rawData.size
            for (i in 7 downTo 0) {
                bits[bitPos++] = ((len shr i) and 1) == 1
            }

            // Data Bits
            for (b in rawData) {
                val bInt = b.toInt()
                for (i in 7 downTo 0) {
                    bits[bitPos++] = ((bInt shr i) and 1) == 1
                }
            }

            // Terminator (Up to 4 zeros)
            val terminatorLimit = min(4, 512 - bitPos)
            for (i in 0 until terminatorLimit) {
                bits[bitPos++] = false
            }

            // Byte Alignment Padding
            while (bitPos % 8 != 0) {
                bits[bitPos++] = false
            }

            // Padding Bytes (0xEC and 0x11)
            var padSwitch = true
            while (bitPos < 512) {
                val padByte = if (padSwitch) 0xEC else 0x11
                for (i in 7 downTo 0) {
                    bits[bitPos++] = ((padByte shr i) and 1) == 1
                }
                padSwitch = !padSwitch
            }

            // Extract Bytes
            val dataWords = IntArray(64)
            for (i in 0 until 64) {
                var b = 0
                for (j in 0 until 8) {
                    if (bits[i * 8 + j]) b = b or (1 shl (7 - j))
                }
                dataWords[i] = b
            }

            // Split into 2 blocks for Version 4-M
            val block1Data = IntArray(32)
            val block2Data = IntArray(32)
            System.arraycopy(dataWords, 0, block1Data, 0, 32)
            System.arraycopy(dataWords, 32, block2Data, 0, 32)

            val gen = getGenerator(18)
            val block1Ecc = calculateEcc(block1Data, gen)
            val block2Ecc = calculateEcc(block2Data, gen)

            // Interleave Data and ECC
            val finalData = IntArray(100)
            var idx = 0
            for (i in 0 until 32) {
                finalData[idx++] = block1Data[i]
                finalData[idx++] = block2Data[i]
            }
            for (i in 0 until 18) {
                finalData[idx++] = block1Ecc[i]
                finalData[idx++] = block2Ecc[i]
            }

            val matrix = Array(33) { IntArray(33) { -1 } }

            // Patterns
            drawFinder(matrix, 0, 0)
            drawFinder(matrix, 26, 0)
            drawFinder(matrix, 0, 26)

            // Alignment Pattern for Version 4 at (26, 26)
            for (y in -2..2) {
                for (x in -2..2) {
                    val isBlack = x == -2 || x == 2 || y == -2 || y == 2 || (x == 0 && y == 0)
                    matrix[26 + y][26 + x] = if (isBlack) 1 else 0
                }
            }

            // Timing Patterns
            for (i in 8 until 25) {
                matrix[6][i] = if (i % 2 == 0) 1 else 0
                matrix[i][6] = if (i % 2 == 0) 1 else 0
            }

            // Dark Module
            matrix[25][8] = 1

            // Reserve Format Info Area
            for (i in 0 until 9) {
                if (matrix[8][i] == -1) matrix[8][i] = 2
                if (matrix[i][8] == -1) matrix[i][8] = 2
            }
            for (i in 25 until 33) {
                if (matrix[8][i] == -1) matrix[8][i] = 2
                if (matrix[i][8] == -1) matrix[i][8] = 2
            }

            // Data Placement (ZigZag)
            var bitIdx = 0
            var dir = -1
            var x = 32
            var y = 32
            while (x > 0) {
                if (x == 6) x--
                while (y in 0 until 33) {
                    for (i in 0 until 2) {
                        val cx = x - i
                        if (matrix[y][cx] == -1) {
                            var bit = false
                            if (bitIdx < 800) {
                                bit = ((finalData[bitIdx / 8] shr (7 - (bitIdx % 8))) and 1) == 1
                            }
                            val maskBit = (y + cx) % 2 == 0
                            matrix[y][cx] = if (bit xor maskBit) 1 else 0
                            bitIdx++
                        }
                    }
                    y += dir
                }
                dir = -dir
                y += dir
                x -= 2
            }

            // Format Info Bits
            val formatBits = intArrayOf(0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1)
            val xs1 = intArrayOf(8, 8, 8, 8, 8, 8, 8, 8, 7, 5, 4, 3, 2, 1, 0)
            val ys1 = intArrayOf(0, 1, 2, 3, 4, 5, 7, 8, 8, 8, 8, 8, 8, 8, 8)
            val xs2 = intArrayOf(8, 8, 8, 8, 8, 8, 8, 25, 26, 27, 28, 29, 30, 31, 32)
            val ys2 = intArrayOf(32, 31, 30, 29, 28, 27, 26, 8, 8, 8, 8, 8, 8, 8, 8)

            for (i in 0 until 15) {
                matrix[ys1[i]][xs1[i]] = formatBits[i]
                matrix[ys2[i]][xs2[i]] = formatBits[i]
            }

            // Pixel Rendering
            val margin = 4
            val matrixSize = 33
            val totalModules = matrixSize + (2 * margin)

            val scale = max(1, tSize / totalModules)
            val finalPixelSize = scale * totalModules

            val pixels = IntArray(finalPixelSize * finalPixelSize)
            for (py in 0 until finalPixelSize) {
                for (px in 0 until finalPixelSize) {
                    val moduleY = (py / scale) - margin
                    val moduleX = (px / scale) - margin

                    var isBlack = false
                    if (moduleY in 0 until matrixSize && moduleX in 0 until matrixSize) {
                        if (matrix[moduleY][moduleX] == 1) {
                            isBlack = true
                        }
                    }
                    pixels[py * finalPixelSize + px] = if (isBlack) Color.BLACK else Color.WHITE
                }
            }

            return Bitmap.createBitmap(finalPixelSize, finalPixelSize, Bitmap.Config.ARGB_8888).apply {
                setPixels(pixels, 0, finalPixelSize, 0, 0, finalPixelSize, finalPixelSize)
            }
        }
    }
}